### PR TITLE
[webkitscmpy] Adopt webkitbugspy for GitHub issues (Part 1)

### PR DESCRIPTION
**Pull request by JonWBedard**
#### 1f96f151ddc2f8936d0b46bb5fab144ae49f5d6d
<pre>
[webkitscmpy] Adopt webkitbugspy for GitHub issues (Part 3)
<a href="https://bugs.webkit.org/show_bug.cgi?id=235899">https://bugs.webkit.org/show_bug.cgi?id=235899</a>
&lt;rdar://problem/88280520 &gt;

Reviewed by Dewei Zhu.

* Tools/Scripts/webkitpy/__init__.py: Add webkitbugspy as a local dependency,
bump local dependency version dependencies.

Canonical link: https://commits.webkit.org/246935@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@289251 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>
----------------------------------------------------------------------
#### 4321c674bb8cf748c4f8f25e8fa9f9a9c7178097
<pre>
[webkitscmpy] Adopt webkitbugspy for GitHub issues (Part 2)
<a href="https://bugs.webkit.org/show_bug.cgi?id=235899">https://bugs.webkit.org/show_bug.cgi?id=235899</a>
&lt;rdar://problem/88280520 &gt;

Reviewed by Dewei Zhu.

 * Tools/Scripts/libraries/webkitscmpy/setup.py: Add webkitbugspy dependency.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
(GitHub): Build on top of webkitbugspy's mock GitHub server.
(GitHub.__init__): Pass mock environment to base class.
(GitHub.request): Let webkitbugspy handle issues.
(GitHub.__enter__): Deleted.
(GitHub.__exit__): Deleted.
(GitHub._users): Deleted.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/git_hub.py:
(GitHub.PRGenerator.PullRequest): Linked issue should be an "issue" object.
(GitHub.PRGenerator.create): Ditto.
(GitHub.PRGenerator.update): Ditto.
(GitHub.PRGenerator._contributor): Leverage webkitbugpy to retrieve information
about unknown contributors.
(GitHub.PRGenerator.comment): Leverage webkitbugspy to create comments.
(GitHub.PRGenerator.comments): Leverage webkitbugspy to retrieve comments.
(GitHub.__init__): Construct a tracker, populate it's user mapping.
(GitHub.credentials): Leverage webkitbugspy to retrieve credentials.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/land_unittest.py:
(TestLandGitHub.webserver): Create a mock issue.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
(TestNetworkPullRequestGitHub.webserver): Create a mock issue.

Canonical link: https://commits.webkit.org/246934@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@289250 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>
----------------------------------------------------------------------
#### 84df1f06fe2fc34017bae685570d9f87b9803ec2
<pre>
[webkitscmpy] Adopt webkitbugspy for GitHub issues (Part 1)
<a href="https://bugs.webkit.org/show_bug.cgi?id=235899">https://bugs.webkit.org/show_bug.cgi?id=235899</a>
&lt;rdar://problem/88280520 &gt;

Reviewed by Dewei Zhu.

* Tools/Scripts/libraries/webkitbugspy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
(Tracker.Encoder.default): Handle API url regexes.
(Tracker.__init__): Match API urls.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/github.py:
(GitHub.__init__): Spell &quot;username&quot; correctly.
(GitHub._issue): Handle case where assignee or watchers is undefined.
(GitHub.request): Rename &quot;api_host&quot; to &quot;api_remote&quot; to match webkitscmpy.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/github_unittest.py:
(TestGitHub.test_link):

Canonical link: <a href="https://commits.webkit.org/246933@main">https://commits.webkit.org/246933@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@289249">https://svn.webkit.org/repository/webkit/trunk@289249</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

**Review by facetothefate — COMMENTED** [review]
LGTM, but need a rubber stamp

File: Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py [hunk 1]
```diff
@@ -47,8 +47,8 @@ def default(context, obj):
                     type='github',
                     url=obj.url,
                 )
-                if obj._res[2:]:
-                    result['res'] = [compiled.pattern for compiled in obj._res[2:]]
+                if obj._res[4:]:
```
Comment (dewei-zhu) [inline Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py]: I wonder if we could make it more dynamic, e.g. self._default_res = [n default regex], self._res = self._default_res + (res or []), the we we do `obj._res[len(self._default_res):]:`

Or create
```
@property

def _res(self):
   return self._default_res + self._extra_res

def __init__(...):
    self._default_res = [...]
    self._extra_res = res or []


if obj._extra_res:
   ...
``

**Review by dewei-zhu — APPROVED** [review]
r=me with comment.

File: Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py (hunk 1 above)
Comment (JonWBedard) [reply Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py]: The `Bugzilla` object has a very similar problem, so I ended up with a slight variation on your first suggestion. The base template regexes were ugly in the constructor as is, so I stuck their templates in a list belonging to the class declaration, now we can just add to the list and we won't have to hard-code lengths anywhere.

**Comment by JonWBedard** [comment]
Landed https://commits.webkit.org/246933@main (84df1f06fe2fc34017bae685570d9f87b9803ec2)!

**Pull request merged**